llvm/lib/Transforms/Vectorize/VPlan.h

```diff
@@ -2144,16 +2151,67 @@ class VPReductionRecipe : public VPSingleDefRecipe {
              VPSlotTracker &SlotTracker) const override;
 #endif
 
+  /// Return the recurrence decriptor for the in-loop reduction.
+  const RecurrenceDescriptor &getRecurrenceDescriptor() const {
+    return RdxDesc;
+  }
+  /// Return true if the in-loop reduction is ordered.
+  bool isOrdered() const { return IsOrdered; };
   /// The VPValue of the scalar Chain being accumulated.
   VPValue *getChainOp() const { return getOperand(0); }
   /// The VPValue of the vector value to be reduced.
   VPValue *getVecOp() const { return getOperand(1); }
   /// The VPValue of the condition for the block.
-  VPValue *getCondOp() const {
+  virtual VPValue *getCondOp() const {
```

Comment:
It's challenging for `VPReduction*`, I think. That's because not all reduction recipes have a `CondOp`. If I put the EVL as *the last* operand, it would be operand(3) if there's a `CondOp`, and operand(2) if there isn't no `CondOp`. Overall, it would be difficult to statically decide whether EVL is operand(2) or operand(3).

Reply:
I don't think there's an easy way around the fact that mask is optional, similar to load/store EVL recipes.

`VPReductionRecipe::classof` likely needs a manual implementation so that isa<VPReductionRecipe>() works for VPReductionEVLRecipe (around like 2143, unfortunately GH doesn't allow commenting on unchanged lines)

Reply:
bdafefa949d5c466007a264c6ff39f8d9098a4c8
62c8979fe81569444ed8a1c4ee51af80295265fe
That's a good reminder.
Although I've refactored some code, to align with the load/store EVL recipes, I'm still keeping the condition as the last operand so far.